src/Microsoft.DotNet.Interactive/Server/NotebookParserServer.cs

```diff
@@ -0,0 +1,409 @@
+                        await Output.FlushAsync();
+                    }
+                }
+            }
+        }
+
+        public static NotebookParserServerResponse HandleRequest(NotebookParseOrSerializeRequest request)
+        {
+            try
+            {
+                switch (request)
+                {
+                    case NotebookParseRequest parse:
+                        {
+                            var contentStream = new MemoryStream(parse.RawData);
+                            var document = request.Extension.ToLowerInvariant() switch
+                            {
+                                ".dib" => CodeSubmission.Read(contentStream, request.DefaultLanguage, WellKnownKernelNames),
+                                ".ipynb" => Notebook.Read(contentStream, WellKnownKernelNames),
+                                _ => throw new NotSupportedException($"Unable to parse an interactive document with extension '{request.Extension}'"),
+                            };
+                            return new NotebookParseResponse(request.Id, document);
+                        }
+                    case NotebookSerializeRequest serialize:
+                        {
+                            var resultStream = new MemoryStream();
+                            switch (request.Extension.ToLowerInvariant())
+                            {
+                                case ".dib":
+                                    CodeSubmission.Write(serialize.Document, serialize.NewLine, resultStream);
+                                    break;
+                                case ".ipynb":
+                                    Notebook.Write(serialize.Document, serialize.NewLine, resultStream);
+                                    break;
+                                default:
+                                    throw new NotSupportedException($"Unable to serialize a interactive document of type '{request.Extension}'");
+                            }
+
+                            resultStream.Position = 0;
+                            var resultArray = resultStream.ToArray();
+                            return new NotebookSerializeResponse(request.Id, resultArray);
+                        }
+                    default:
+                        return new NotebookErrorResponse(request.Id, $"Unsupported request: {request}");
+                }
+            }
+            catch (Exception ex)
+            {
+                return new NotebookErrorResponse(request.Id, ex.ToString());
+            }
+        }
+    }
+
+    public abstract class NotebookParseOrSerializeRequest
+    {
+        public string Type { get; set; }
+        public string Id { get; set; }
+        public string Extension { get; set; }
+        public string DefaultLanguage { get; set; }
+
+        protected NotebookParseOrSerializeRequest(string type, string id, string extension, string defaultLanguage)
```

Comment:
What's the point of the constructor if the properties have setters?

Reply:
This is the result of me fighting with the JSON deserializer and trying to make it happy.  I should be able to get rid of the setters.